### PR TITLE
Ignore Deck Movement Config Settings Option

### PR DESCRIPTION
**Pull request by ohare93**
Can't stop won't stop :musical_note: Fix for Issue #23 :+1: 

Had to pass config around in a lot of places :sweat_smile: it may have just been easier to call the static version where I needed it, but I imagine this work will be used later on as well for other such config options.

All tests passing, have manually tested import/export without the feature, and the feature itself works for me :grin:

### CODE REVIEW

File: crowd_anki/representation/note.py [hunk 1]
```diff
@@ -117,5 +123,5 @@ def save_to_collection(self, collection, deck, model_map_cache):
 
         if new_note:
             collection.addNote(self.anki_object)
-        else:
+        elif not self.config.import_notes_ignore_deck_movement:
```
Comment (Stvad) [inline crowd_anki/representation/note.py]: there is too much of passing things around, I'd prefer if you use a pattern similar to what we've done with `addonManager` here. (making the config one of the `__init__` parameters with None as default value and initializing it via singleton value if nothing is passed in)

File: crowd_anki/config/config_dialog.py [hunk 2]
```diff
@@ -32,6 +32,7 @@ def accept(self):
             super().accept()
 
     def ui_initial_setup(self):
+        # Snapshot Options
```
Comment (Stvad) [inline crowd_anki/config/config_dialog.py]: better just extract each of those sections into a separate method

File: ui_files/config.ui [hunk 3]
```diff
@@ -58,6 +58,22 @@
...
+       <item>
+        <widget class="QGroupBox" name="group_deck_import">
+         <property name="title">
+          <string>Import</string>
+         </property>
+         <layout class="QVBoxLayout" name="verticalLayout_5">
+          <item>
+           <widget class="QCheckBox" name="cb_ignore_move_cards">
+            <property name="text">
+             <string>Do Not Move Existing Cards</string>
```
Comment (Stvad) [inline ui_files/config.ui]: I think this warrant a more detailed explanation, potentially with a reference to a github issue

File: crowd_anki/representation/note.py (hunk 1 above)
Comment (ohare93) [reply crowd_anki/representation/note.py]: Yea thought you may say that, no worries. I am inclined to agree :sweat_smile: Done :+1:

File: crowd_anki/config/config_dialog.py (hunk 2 above)
Comment (ohare93) [reply crowd_anki/config/config_dialog.py]: Done :+1:

File: ui_files/config.ui (hunk 3 above)
Comment (ohare93) [reply ui_files/config.ui]: Done :+1: give it a look

File: crowd_anki/config/config_dialog.py [hunk 4]
```diff
@@ -32,6 +32,11 @@ def accept(self):
             super().accept()
 
     def ui_initial_setup(self):
+        self.ui_setup_snapshot_options()
```
Comment (Stvad) [inline crowd_anki/config/config_dialog.py]: I don't think we need the `ui` in the name, we're in the ui widget after all ;).

File: crowd_anki/representation/json_serializable.py [hunk 5]
```diff
@@ -115,7 +115,7 @@ def from_json(cls, json_dict):
 
 
 class JsonSerializableAnkiObject(JsonSerializable):
-    export_filter_set = JsonSerializable.export_filter_set | {"anki_object", "anki_object_dict"}
```
Comment (Stvad) [inline crowd_anki/representation/json_serializable.py]: this should be in the `Note` filter

File: crowd_anki/representation/note.py [hunk 6]
```diff
@@ -38,8 +41,8 @@ def from_collection(cls, collection, note_id, note_models):
         return note
 
     @classmethod
-    def from_json(cls, json_dict):
-        note = Note()
+    def from_json(cls, json_dict, config: ConfigSettings = None):
+        note = Note(config=config)
```
Comment (Stvad) [inline crowd_anki/representation/note.py]: why do we need these?

File: crowd_anki/representation/json_serializable.py (hunk 5 above)
Comment (ohare93) [reply crowd_anki/representation/json_serializable.py]: Okay, done :+1:

File: crowd_anki/representation/note.py (hunk 6 above)
Comment (ohare93) [reply crowd_anki/representation/note.py]: Well I added an optional config parameter to the Note constructor, and these are the only functions which ever call the Note constructor. They are used throughout the code, as the default way to make notes, so I thought they also needed the ability to have this config item passed to them :thinking: No one is ever going to simple call Note()

File: crowd_anki/config/config_dialog.py (hunk 4 above)
Comment (ohare93) [reply crowd_anki/config/config_dialog.py]: Yea sure why not :ok:

File: crowd_anki/representation/note.py (hunk 6 above)
Comment (Stvad) [reply crowd_anki/representation/note.py]: eh, this argument have some merit, but I'm more of a mind that the code that is not used should not be there. Do you envision this being used any time soon?

Comment (ohare93) [reply crowd_anki/representation/note.py]: Nope, not in the slightest! I shall remove it then :wink:

File: crowd_anki/representation/note.py [hunk 7]
```diff
@@ -28,7 +32,7 @@ def get_notes_from_collection(collection, deck_id, note_models):
     @classmethod
     def from_collection(cls, collection, note_id, note_models):
         anki_note = AnkiNote(collection, id=note_id)
-        note = Note(anki_note)
+        note = Note(anki_note=anki_note)
```
Comment (Stvad) [inline crowd_anki/representation/note.py]: 👀

Comment (ohare93) [reply crowd_anki/representation/note.py]: :thinking: ?

The Note constructor now has two optional parameters:

```python
    def __init__(self, anki_note=None, config: ConfigSettings = None):
        super(Note, self).__init__(anki_note)
        self.note_model_uuid = None
        self.config = config or ConfigSettings.get_instance()
```

So I made sure to be explicit about which optional parameter:
```python
    @classmethod
    def from_collection(cls, collection, note_id, note_models):
        anki_note = AnkiNote(collection, id=note_id)
        note = Note(anki_note=anki_note)
```

Comment (Stvad) [reply crowd_anki/representation/note.py]: they are still positional though 🤷‍♂️

Comment (ohare93) [reply crowd_anki/representation/note.py]: _Readable and explicit!_ 😁 🤷‍♂

**Comment by ohare93** [comment]
Pulled already? That was easy 😁 👍

**Comment by Stvad** [comment]
yep, thank you for you contributions ;)

**Pull request merged**